File: leetcode/python/116. 填充每个节点的下一个右侧节点指针.py
#!/bin/bash python
"""
116. 填充每个节点的下一个右侧节点指针

思路一：
bfs层次优先遍历

思路二：
dfs
"""

class Node(object):
    def __init__(self, val=0, left=None, right=None, next=None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next

class Solution(object):
    def connect(self, root):
        """
        层次优先法，队列
        """
        if root == None:
            return root
        queue = [root]
        while queue:
            size = len(queue)
            for i in range(size):
                node = queue.pop(0)
                node.next = queue[0] if queue and i != size - 1 else None
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
        return root
    
    def connect2(self, root):
        """
        利用next指针的先序遍历
        """
        def dfs(root, nextNode):
            """
            当前root的下一个节点为nextNode
            遍历当前树，并且构建每个节点的nextNode
            """
            if root == None:
                return
            root.next = nextNode
            dfs(root.left, root.right)
            dfs(root.right, None if root.next == None else root.next.left)
            return
        dfs(root, None)
        return root

File: leetcode/cpp/116. 填充每个节点的下一个右侧节点指针.cpp
#include <stdio.h>
#include <queue>

typedef struct Node {
    int val;
    Node* left;
    Node* right;
    Node* next;
    Node(int val, Node* left, Node* right, Node* next) : 
        val(val), left{left}, right(right), next(next) {}
} Node;

class Solution {
public:
    // 层次遍历法
    Node* connect(Node* root) {
        if (root == nullptr) {
            return root;
        }
        std::queue<Node*> queue;
        queue.push(root);
        while (!queue.empty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                Node* node = queue.front();
                queue.pop();
                if (!queue.empty() && i != size - 1) {
                    node->next = queue.front();
                }
                if (node->left) {
                    queue.push(node->left);
                }
                if (node->right) {
                    queue.push(node->right);
                }
            }
        }
        return root;
    }

    // 带下一个节点的先序遍历
    void dfs(Node* root, Node* nextNode) {
        if (root == nullptr) {
            return;
        }
        root->next = nextNode;
        dfs(root->left, root->right);
        dfs(root->right, nextNode ? nextNode->left : nullptr);
        return;
    }

    // 直接先序遍历
    Node* connect2(Node* root) {
        dfs(root, nullptr);
        return root;
    }
};

int main(int argc, char* argv[])
{
    return 0;
}